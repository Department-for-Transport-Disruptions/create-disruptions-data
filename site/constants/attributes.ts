export const CD_SUMMARY = "summary";

export const CD_DESCRIPTION = "description";

export const CD_DISRUPTION_TYPE = "disruptionType";

export const CD_ASSOCIATED_LINK = "associatedLink";

export const CD_DISRUPTION_REASON = "disruptionReason";

export const CD_DISRUPTION_REAPEATS = "disruptionRepeats";

export const CD_NO_VALIDITY_END_DATE_TIME = "validityNoEndDateTime";

export const CD_NO_PUBLISH_END_DATE_TIME = "publishNoEndDateTime";

export const CD_VALIDITY_START_DATE_DAY = "validityStartDateDay";

export const CD_VALIDITY_START_DATE_MONTH = "validityStartDateMonth";

export const CD_VALIDITY_START_DATE_YEAR = "validityStartDateYear";

export const CD_VALIDITY_START_HOUR = "validityStartHour";

export const CD_VALIDITY_START_MINUTES = "validityStartMinute";

export const CD_VALIDITY_END_DATE_DAY = "validityEndDateDay";

export const CD_VALIDITY_END_DATE_MONTH = "validityEndDateMonth";

export const CD_VALIDITY_END_DATE_YEAR = "validityEndDateYear";

export const CD_VALIDITY_END_HOUR = "validityEndHour";

export const CD_VALIDITY_END_MINUTES = "validityEndMinute";

export const CD_PUBLISH_START_DATE_DAY = "publishStartDateDay";

export const CD_PUBLISH_START_DATE_MONTH = "publishStartDateMonth";

export const CD_PUBLISH_START_DATE_YEAR = "publishStartDateYear";

export const CD_PUBLISH_START_HOUR = "publishStartHour";

export const CD_PUBLISH_START_MINUTES = "publishStartMinute";

export const CD_PUBLISH_END_DATE_DAY = "publishEndDateDay";

export const CD_PUBLISH_END_DATE_MONTH = "publishEndDateMonth";

export const CD_PUBLISH_END_DATE_YEAR = "publishEndDateYear";

export const CD_PUBLISH_END_HOUR = "publishEndHour";

export const CD_PUBLISH_END_MINUTES = "publishEndMinute";
