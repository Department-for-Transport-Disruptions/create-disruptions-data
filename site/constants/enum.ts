export enum TransportMode {
    bus = "Bus",
    tram = "Tram",
    ferry = "Ferry",
    train = "Train",
}

export enum ConsequenceType {
    services = "Services",
    networkWide = "Network wide",
    operatorWide = "Operator wide",
    stops = "Stops",
}
