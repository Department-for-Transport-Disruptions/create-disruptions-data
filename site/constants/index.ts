import {
    MiscellaneousReason,
    PersonnelReason,
    EnvironmentReason,
    EquipmentReason,
} from "@create-disruptions-data/shared-ts/enums";

export const SUPPORT_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SERVICE_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SUPPORT_PHONE_NUMBER = "0800 123 1234";

export const DISRUPTION_REASONS: {
    value: MiscellaneousReason | PersonnelReason | EnvironmentReason | EquipmentReason;
    display: string;
}[] = [
    {
        value: MiscellaneousReason.roadWorks,
        display: "Road Works",
    },
    {
        value: MiscellaneousReason.vandalism,
        display: "Vandalism",
    },
    {
        value: MiscellaneousReason.routeDiversion,
        display: "Route Diversion",
    },
    {
        value: MiscellaneousReason.specialEvent,
        display: "Special Event",
    },
];

export const DISRUPTION_TYPES = ["planned", "unplanned"];

export const CD_DATE_FORMAT = "DD/MM/YYYY";
export const CONTACT_FEEDBACK_QUESTION = "Did you contact us for assistance at any point?";
export const SOLVE_FEEDBACK_QUESTION = "Did we solve your problem?";
export const HEAR_ABOUT_US_FEEDBACK_QUESTION = "How did you hear about our service?";
export const GENERAL_FEEDBACK_QUESTION = "Please let us know any feedback or suggestions for improvements you may have";

export const COOKIES_POLICY_COOKIE = "cdd-cookies-policy";

export const COOKIE_PREFERENCES_COOKIE = "cdd-cookie-preferences-set";

export const oneYearInSeconds = 31556952;

export const ID_TOKEN_COOKIE = "cdd-id-token";

export const COOKIES_DISRUPTION_INFO = "disruption-info";

export const COOKIES_DISRUPTION_ERRORS = "disruption-errors";

export const COOKIES_ADD_CONSEQUENCE_INFO = "cdd-consequence-info";

export const COOKIES_ADD_CONSEQUENCE_ERRORS = "cdd-consequence-errors";

export const CREATE_DISRUPTION_PAGE_PATH = "/create-disruption";

export const ERROR_PATH = "/_error";

export const ADD_CONSEQUENCE_PAGE_PATH = "/type-of-consequence";

export const STAGE = process.env.STAGE || "dev";
