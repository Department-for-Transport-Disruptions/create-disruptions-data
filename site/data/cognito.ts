import {
    CognitoIdentityProviderClient,
    AdminInitiateAuthCommand,
    AdminInitiateAuthCommandInput,
    AdminInitiateAuthCommandOutput,
    AdminRespondToAuthChallengeCommand,
    AdminRespondToAuthChallengeCommandInput,
    AdminUserGlobalSignOutCommand,
    AdminUserGlobalSignOutCommandInput,
    AdminSetUserPasswordCommand,
    AdminSetUserPasswordCommandInput,
    AdminAddUserToGroupCommand,
    ListGroupsCommand,
    ListUsersInGroupCommand,
    UserType,
    AdminDeleteUserCommand,
    AdminGetUserCommand,
    AdminGetUserCommandInput,
    AdminDeleteUserCommandInput,
    AdminCreateUserCommand,
    ListUsersCommand,
    ConfirmForgotPasswordCommand,
    ConfirmForgotPasswordCommandInput,
    ForgotPasswordCommandInput,
    ForgotPasswordCommand,
} from "@aws-sdk/client-cognito-identity-provider";

import { createHmac } from "crypto";
import { AddUserSchema } from "../schemas/add-user.schema";
import { userManagementSchema } from "../schemas/user-management.schema";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const {
    COGNITO_CLIENT_ID: cognitoClientId,
    COGNITO_CLIENT_SECRET: cognitoClientSecret,
    COGNITO_USER_POOL_ID: userPoolId,
} = process.env;

if (!cognitoClientSecret || !cognitoClientId || !userPoolId) {
    throw new Error("Cognito env vars not set");
}

const cognito = new CognitoIdentityProviderClient({
    region: "eu-west-2",
});

export const deleteUser = (username: string) => {
    try {
        const params: AdminDeleteUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminDeleteUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to delete user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getUserDetails = (username: string) => {
    try {
        const params: AdminGetUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminGetUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to get user details by username: ${error.stack || ""}`);
        }

        throw error;
    }
};

const calculateSecretHash = (username: string): string =>
    createHmac("SHA256", cognitoClientSecret)
        .update(username + cognitoClientId)
        .digest("base64");

export const initiateAuth = async (username: string, password: string): Promise<AdminInitiateAuthCommandOutput> => {
    logger.info("", {
        context: "data.cognito",
        message: "initiating auth",
    });

    try {
        const params: AdminInitiateAuthCommandInput = {
            AuthFlow: "ADMIN_USER_PASSWORD_AUTH",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            AuthParameters: {
                USERNAME: username,
                PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
        };

        return cognito.send(new AdminInitiateAuthCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to authenticate user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "new password challenge initiated",
    });

    try {
        const params: AdminRespondToAuthChallengeCommandInput = {
            ChallengeName: "NEW_PASSWORD_REQUIRED",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            ChallengeResponses: {
                USERNAME: username,
                NEW_PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
            Session: session,
        };

        await cognito.send(new AdminRespondToAuthChallengeCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to respond to password challenge: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "performing global sign out",
    });

    const params: AdminUserGlobalSignOutCommandInput = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminUserGlobalSignOutCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to perform global sign out: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "updating user password",
    });

    const params: AdminSetUserPasswordCommandInput = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminSetUserPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to update user password: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getAllUsersInGroup = async (groupName: string, nextToken?: string): Promise<UserType[]> => {
    const users = await cognito.send(
        new ListUsersInGroupCommand({ UserPoolId: userPoolId, GroupName: groupName, Limit: 60, NextToken: nextToken }),
    );

    if (!users.Users) {
        return [];
    }

    if (users.NextToken) {
        return [...users.Users, ...(await getAllUsersInGroup(groupName, users.NextToken))];
    } else {
        return users.Users;
    }
};

export const listUsersWithGroups = async () => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users",
    });

    try {
        const groupList = await cognito.send(new ListGroupsCommand({ UserPoolId: userPoolId }));

        const userList = await Promise.all(
            groupList.Groups?.map(async (group) => {
                if (!group.GroupName) {
                    return null;
                }

                const groupUsers = await getAllUsersInGroup(group.GroupName);

                return groupUsers.map((user) => ({ ...user, GroupName: group.GroupName }));
            }) ?? [],
        );

        const parsedUsers = userManagementSchema.safeParse(userList.filter(notEmpty).flat());

        if (!parsedUsers.success) {
            return [];
        }

        return parsedUsers.data;
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to list cognito users: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const deleteUsersByAttribute = async (attributeName: string, attributeValue: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users based on custom attributes and deleting the users that match the value",
    });

    try {
        const userList = await cognito.send(
            new ListUsersCommand({ UserPoolId: userPoolId, AttributesToGet: [attributeName] }),
        );

        const deletePromises = userList.Users?.filter((user) =>
            user.Attributes?.find((value) => value.Value === attributeValue),
        ).map((user) => {
            return cognito.send(
                new AdminDeleteUserCommand({
                    UserPoolId: userPoolId,
                    Username: user.Username,
                }),
            );
        });

        await Promise.all(deletePromises ?? []);
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(
                `Failed to list cognito users based on attribute ${attributeName} with value ${attributeValue}: ${
                    error.stack || ""
                }`,
            );
        }

        throw error;
    }
};

export const createUser = async (userData: AddUserSchema) => {
    logger.info("", {
        context: "data.cognito",
        message: "Adding a new user",
    });

    const nocCodes = userData.operatorNocInfo?.map((operator) => operator.nocCode) ?? [];

    const userAttributes = [
        {
            Name: "custom:orgId",
            Value: userData.orgId,
        },
        {
            Name: "given_name",
            Value: userData.givenName,
        },
        {
            Name: "family_name",
            Value: userData.familyName,
        },
        {
            Name: "email_verified",
            Value: "true",
        },
        {
            Name: "email",
            Value: userData.email,
        },
    ];

    const createUserCommand = {
        Username: userData.email,
        UserPoolId: userPoolId,
        TemporaryPassword: Array.from(Array(20), () => Math.floor(Math.random() * 36).toString(36)).join(""),
        UserAttributes: userAttributes,
    };

    const createOperatorUserCommand = {
        ...createUserCommand,
        UserAttributes: userAttributes.concat([
            {
                Name: "custom:nocCodes",
                Value: nocCodes.toString(),
            },
        ]),
    };

    const createUserResult =
        userData.group === "operators"
            ? await cognito.send(new AdminCreateUserCommand(createOperatorUserCommand))
            : await cognito.send(new AdminCreateUserCommand(createUserCommand));

    await cognito.send(
        new AdminAddUserToGroupCommand({
            GroupName: userData.group,
            Username: createUserResult.User?.Username,
            UserPoolId: userPoolId,
        }),
    );
};

export const initiateResetPassword = async (email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Initiating reset password flow for cognito user",
    });

    try {
        const params: ForgotPasswordCommandInput = {
            Username: email,
            ClientId: cognitoClientId,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to initiate reset password flow: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const resetUserPassword = async (key: string, newPassword: string, email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Resetting password for cognito user",
    });

    try {
        const params: ConfirmForgotPasswordCommandInput = {
            ClientId: cognitoClientId,
            ConfirmationCode: key,
            Password: newPassword,
            Username: email,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ConfirmForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to reset password: ${error.stack || ""}`);
        }

        throw error;
    }
};
