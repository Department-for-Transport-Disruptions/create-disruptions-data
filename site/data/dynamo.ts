import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { ptSituationElementSchema } from "@create-disruptions-data/shared-ts/siriTypes.zod";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({ region: "eu-west-2" }));

const tableName = process.env.TABLE_NAME as string;

export const getDisruptionsDataFromDynamo = async () => {
    logger.info("Getting disruptions data from DynamoDB table...");

    const dbData = await ddbDocClient.send(
        new QueryCommand({
            TableName: tableName,
            KeyConditionExpression: "PK = :i",
            ExpressionAttributeValues: {
                ":i": "1",
            },
        }),
    );

    return dbData.Items?.map((item) => {
        const parsedItem = ptSituationElementSchema.safeParse(item);

        if (!parsedItem.success) {
            logger.error("Error parsing disruption from dynamo");
            logger.error(parsedItem.error.stack);
            return null;
        }

        return parsedItem.data;
    }).filter(notEmpty);
};
