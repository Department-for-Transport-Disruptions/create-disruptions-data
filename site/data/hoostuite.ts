import startCase from "lodash/startCase";
import { NextPageContext } from "next";
import { parseCookies } from "nookies";
import { getParameter, getParametersByPath, putParameter } from "./ssm";
import { COOKIES_ID_TOKEN, COOKIES_REFRESH_TOKEN, HOOTSUITE_URL } from "../constants";
import { hootsuiteMeSchema, hootsuiteTokenSchema, hootsuiteSocialProfilesSchema } from "../schemas/hootsuite.schema";
import { HootsuiteProfiles, SocialMediaAccountsSchema } from "../schemas/social-media-accounts.schema";

export const getHootsuiteToken = async (refreshToken: string, authToken: string) => {
    return await fetch(`${HOOTSUITE_URL}oauth2/token`, {
        method: "POST",
        body: new URLSearchParams({
            grant_type: "refresh_token",
            refresh_token: refreshToken ?? "",
        }),
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            Authorization: authToken,
        },
    });
};

export const getHootsuiteData = async (
    ctx: NextPageContext,
    username: string,
    orgId: string,
): Promise<{ clientId: string; userData: SocialMediaAccountsSchema }> => {
    const cookies = parseCookies(ctx);
    let clientIdValue = "";
    let userData: SocialMediaAccountsSchema = [];
    try {
        const idToken = cookies[COOKIES_ID_TOKEN];
        const refreshToken = cookies[COOKIES_REFRESH_TOKEN];

        const [clientId, clientSecret, keys, tokensByOrganisation] = await Promise.all([
            getParameter(`/social/hootsuite/client_id`),
            getParameter(`/social/hootsuite/client_secret`),
            getParametersByPath(`/social/${orgId}/hootsuite`),
            getParametersByPath(`/social/${orgId}/hootsuite`),
        ]);

        if (!clientId || !clientSecret) {
            throw new Error("clientId and clientSecret must be defined");
        }

        clientIdValue = clientId.Parameter?.Value || "";
        const key = `${clientId.Parameter?.Value || ""}:${clientSecret.Parameter?.Value || ""}`;

        const authToken = `Basic ${Buffer.from(key).toString("base64")}`;

        if (idToken && refreshToken)
            await Promise.all([
                putParameter(`/${username}/token`, idToken, "SecureString", true),
                putParameter(`/${username}/refresh-token`, refreshToken, "SecureString", true),
            ]);

        const refreshTokens = tokensByOrganisation?.Parameters?.map((token) => {
            return {
                value: token.Value,
                name: token.Name,
                userId: token?.Name?.split("hootsuite/")[1]?.split("-")[0] ?? "",
                accountType: startCase(token?.Name?.split("/")[3]) ?? "",
                addedBy: token?.Name?.split("/")[4]?.replace("_", " ")?.split("-")[1] ?? "",
            };
        });

        if (refreshTokens && refreshTokens.length > 0) {
            await Promise.all(
                refreshTokens?.map(async (token) => {
                    const resp = await getHootsuiteToken(token.value || "", authToken);
                    if (resp.ok) {
                        const tokenResult = hootsuiteTokenSchema.parse(await resp.json());

                        if (!keys || (refreshTokens && keys.Parameters?.length === 0)) {
                            throw new Error("Refresh token is required to fetch dropdown data");
                        }
                        const key: string =
                            keys.Parameters?.find((rt) => rt.Name?.includes(`${token.userId}`))?.Name || "";
                        if (!key) {
                            throw new Error("Refresh token is required to fetch dropdown data");
                        }

                        await putParameter(key, tokenResult.refresh_token ?? "", "SecureString", true);
                        const userDetailsResponse = await fetch(`${HOOTSUITE_URL}v1/me`, {
                            method: "GET",
                            headers: {
                                Authorization: `Bearer ${tokenResult.access_token ?? ""}`,
                            },
                        });
                        if (userDetailsResponse.ok) {
                            const userDetails = hootsuiteMeSchema.parse(await userDetailsResponse.json());
                            const userInfo = userDetails.data || {};

                            const extraInfo = {
                                ...userInfo,
                                accountType: token.accountType || "",
                                addedBy: token.addedBy || "",
                                expiresIn: "Never",
                            };

                            const socialProfilesResponse = await fetch(`${HOOTSUITE_URL}v1/socialProfiles`, {
                                method: "GET",
                                headers: {
                                    Authorization: `Bearer ${tokenResult.access_token ?? ""}`,
                                },
                            });
                            if (socialProfilesResponse.ok) {
                                const socialProfiles = hootsuiteSocialProfilesSchema.parse(
                                    await socialProfilesResponse.json(),
                                );

                                userData = [
                                    ...userData,
                                    {
                                        ...extraInfo,
                                        hootsuiteProfiles:
                                            socialProfiles.data?.map((sp: HootsuiteProfiles[0]) => ({
                                                type: sp.type,
                                                socialNetworkId: sp.socialNetworkId,
                                                id: sp.id,
                                            })) ?? [],
                                    },
                                ];
                            }
                        }
                    }
                }),
            );
        }

        return { clientId: clientIdValue, userData };
    } catch (e) {
        throw new Error(`${(e as Error).message}`);
    }
};
