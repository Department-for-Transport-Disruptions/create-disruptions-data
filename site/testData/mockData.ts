/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { mockRequest, mockResponse } from "mock-req-res";
import { NextPageContext } from "next";
import React from "react";
import { Mock, vi } from "vitest";
import { COOKIE_ID_TOKEN, COOKIES_POLICY_COOKIE } from "../constants";

export interface GetMockContextInput {
    session?: { [key: string]: any };
    cookies?: any;
    body?: any;
    url?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    isLoggedin?: boolean;
    query?: any;
}

export const getMockRequestAndResponse = ({
    cookieValues = {},
    body = null,
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    requestHeaders = {},
    isLoggedin = true,
    url = null,
    query = null,
}: GetMockRequestAndResponse = {}): { req: any; res: any } => {
    const res = mockResponse({ writeHead: mockWriteHeadFn, end: mockEndFn });

    const {
        idToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJjdXN0b206bm9jIjoiVEVTVCIsImVtYWlsIjoidGVzdEBleGFtcGxlLmNvbSIsImp0aSI6Ijg1MmQ1MTVlLTU5YWUtNDllZi1iMTA5LTI4YTRhNzk3YWFkNSIsImlhdCI6MTU5Mjk4NzMwNywiZXhwIjoxNTkyOTkwOTA3fQ.DFdxnpdhykDONOMeZMNeMUFpCHZ-hQ3UXczq_Qh0IAI",
        cookiePolicy = null,
    } = cookieValues;

    const defaultSession = {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        destroy: (): void => {},
    };

    let cookieString = "";

    cookieString += isLoggedin ? `${COOKIE_ID_TOKEN}=${idToken as string};` : "";

    cookieString += cookiePolicy ? `${COOKIES_POLICY_COOKIE}=${encodeURI(JSON.stringify(cookiePolicy))}` : "";

    const req = mockRequest({
        connection: {
            encrypted: true,
        },
        url,
        headers: {
            host: "localhost:3000",
            cookie: cookieString,
            origin: "localhost:3000",
            ...requestHeaders,
        },
        cookies: cookieValues,
        session: { ...defaultSession },
    });

    if (body) {
        req.body = body;
    }
    if (query) {
        req.query = query;
    }

    return { req, res };
};

export const getMockContext = ({
    cookies = {},
    body = null,
    uuid = {},
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    isLoggedin = true,
    url = null,
    query = "",
}: GetMockContextInput = {}): NextPageContext => {
    const { req, res } = getMockRequestAndResponse({
        cookieValues: cookies,
        body,
        uuid,
        mockWriteHeadFn,
        mockEndFn,
        requestHeaders: {},
        isLoggedin,
        url,
    });

    const ctx: NextPageContext = {
        res,
        req,
        pathname: "",
        query,
        // eslint-disable-next-line react/display-name
        AppTree: () => React.createElement("div"),
    };

    return ctx;
};

export interface GetMockRequestAndResponse {
    cookieValues?: any;
    body?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    requestHeaders?: any;
    isLoggedin?: boolean;
    url?: any;
    query?: any;
}
