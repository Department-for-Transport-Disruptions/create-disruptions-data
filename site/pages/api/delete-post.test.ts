import MockDate from "mockdate";
import { describe, it, expect, afterEach, vi, afterAll } from "vitest";
import deleteSocialMediaPost from "./delete-post.api";
import { DISRUPTION_DETAIL_PAGE_PATH, ERROR_PATH, REVIEW_DISRUPTION_PAGE_PATH } from "../../constants/index";
import * as dynamo from "../../data/dynamo";
import { getMockRequestAndResponse } from "../../testData/mockData";

const defaultDisruptionId = "acde070d-8c4c-4f0d-9d8a-162843c10333";
const defaultSocialMediaPostId = "1";

describe("deletePost", () => {
    const writeHeadMock = vi.fn();
    vi.mock("../../utils/apiUtils", async () => ({
        ...(await vi.importActual<object>("../../utils/apiUtils")),
        setCookieOnResponseObject: vi.fn(),
        destroyCookieOnResponseObject: vi.fn(),
        cleardownCookies: vi.fn(),
    }));

    vi.mock("../../data/dynamo", () => ({
        removeSocialMediaPostFromDisruption: vi.fn(),
        upsertSocialMediaPost: vi.fn(),
    }));

    vi.mock("crypto", () => ({
        randomUUID: () => "id",
    }));

    MockDate.set("2023-03-03");

    const deleteConsequenceSpy = vi.spyOn(dynamo, "removeSocialMediaPostFromDisruption");

    afterEach(() => {
        vi.resetAllMocks();
    });

    afterAll(() => {
        MockDate.reset();
    });

    it("should retrieve valid data from cookies, write to dynamo and redirect", async () => {
        const { req, res } = getMockRequestAndResponse({
            body: {
                id: defaultSocialMediaPostId,
                disruptionId: defaultDisruptionId,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await deleteSocialMediaPost(req, res);

        expect(dynamo.removeSocialMediaPostFromDisruption).toBeCalledTimes(1);
        expect(writeHeadMock).toBeCalledWith(302, {
            Location: `${REVIEW_DISRUPTION_PAGE_PATH}/${defaultDisruptionId}`,
        });
    });

    it("should redirect to /dashboard page after updating consequence when invoked from disruption details page", async () => {
        const { req, res } = getMockRequestAndResponse({
            body: {
                id: defaultSocialMediaPostId,
                disruptionId: defaultDisruptionId,
                inEdit: "true",
            },
            mockWriteHeadFn: writeHeadMock,
        });

        await deleteSocialMediaPost(req, res);

        expect(dynamo.upsertSocialMediaPost).toBeCalledTimes(1);
        expect(writeHeadMock).toBeCalledWith(302, {
            Location: `${DISRUPTION_DETAIL_PAGE_PATH}/${defaultDisruptionId}`,
        });
    });

    it("should redirect to error page if consequenceId not passed", async () => {
        const { req, res } = getMockRequestAndResponse({
            mockWriteHeadFn: writeHeadMock,
        });

        await deleteSocialMediaPost(req, res);

        expect(dynamo.removeSocialMediaPostFromDisruption).not.toBeCalled();
        expect(writeHeadMock).toBeCalledWith(302, { Location: ERROR_PATH });
    });

    it("should delete consequence data in dynamoDB"),
        async () => {
            const { req, res } = getMockRequestAndResponse({
                body: {
                    id: defaultSocialMediaPostId,
                    disruptionId: defaultDisruptionId,
                },
                mockWriteHeadFn: writeHeadMock,
            });

            await deleteSocialMediaPost(req, res);

            expect(deleteConsequenceSpy.mock.calls[0][0]).toMatchSnapshot();
        };
});
