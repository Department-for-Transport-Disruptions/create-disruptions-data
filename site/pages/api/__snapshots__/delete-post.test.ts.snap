// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deletePost > should delete post data in dynamoDB 1`] = `1`;
